### PR TITLE
ZenML Failure Hook for OpenAI ChatGPT fixes

### PR DESCRIPTION
**Pull request by strickvl**
I implemented a failure hook that passes the stacktrace and exception to ChatGPT (`gpt-3.5-turbo` model) to output suggestions on how to fix the failing step. It uses `rich`'s traceback to get more context on the failing code and passes those suggestions through to a registered alerter component. Output looks something like this:

failure_alerter

Users have to pass in their OpenAI API key as a secret. It's a bit clunky, but hard to validate or pass down configuration any other way since this failure hook behaviour isn't represented by a stack component.

## Pre-requisites
Please ensure you have done the following:
- [x] I have read the **CONTRIBUTING.md** document.
- [x] If my change requires a change to docs, I have updated the documentation accordingly.
- [ ] If I have added an integration, I have updated the [integrations](https://docs.zenml.io/component-gallery/integrations) table and the [corresponding website section](https://zenml.io/integrations).
- [ ] I have added tests to cover my changes.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Other (add details above)

### CODE REVIEW

File: src/zenml/integrations/openai/hooks/open_ai_failure_hook.py
Comment (htahir1) [inline src/zenml/integrations/openai/hooks/open_ai_failure_hook.py:50]: Do we need to use console here? Not sure why its neccessary and if it is we should use it in the other standard hook as well i suppose?

File: src/zenml/integrations/openai/hooks/open_ai_failure_hook.py
Comment (htahir1) [inline src/zenml/integrations/openai/hooks/open_ai_failure_hook.py:56]: I wonder if there is a way to parameterize this... not sure there is an easy way

**Review by htahir1 — CHANGES_REQUESTED** [review]
What a beautiful feature! I really love this sort of innovation. Kudos!

Other than my comments, some minor testing might be worth taking a look?

File: src/zenml/integrations/openai/hooks/open_ai_failure_hook.py
Comment (strickvl) [reply src/zenml/integrations/openai/hooks/open_ai_failure_hook.py:50]: It's the rich output console which gives a much more detailed traceback. Currently the standard hook is printing just the exception type and not the traceback. Not sure if we'd want to output the full traceback? In any case, we only need the `rich` console if we're getting that full traceback.

File: src/zenml/integrations/openai/hooks/open_ai_failure_hook.py
Comment (strickvl) [reply src/zenml/integrations/openai/hooks/open_ai_failure_hook.py:56]: I thought about using a `partial` here, but I think no way then to validate the hook function without either ignoring some annotations. I think we can just offer it as an example of how you could do this, and then people can of course make their own.

Perhaps `hook` functions would be in the hub as well at some point? For now, I'll make a standard and a gpt-4 version of the hook.

**Review by safoinme — CHANGES_REQUESTED** [review]

File: docs/book/advanced-guide/pipelines/hooks.md
Comment (safoinme) [inline docs/book/advanced-guide/pipelines/hooks.md:145]:
```suggestion
from zenml.integration.openai.hooks import openai_chatgpt_alerter_failure_hook

@step(on_failure=openai_chatgpt_alerter_failure_hook)
def my_step(...):
```

**Review by safoinme — CHANGES_REQUESTED** [review]

**Review by safoinme — APPROVED** [review]
LGTM

File: src/zenml/integrations/openai/hooks/open_ai_failure_hook.py
Comment (htahir1) [reply src/zenml/integrations/openai/hooks/open_ai_failure_hook.py:47]: Id say we should be consistent, so if we do this here, then lets update the other one

**Review by htahir1 — COMMENTED** [review]
Apart from my one comment above, how about adding in these blocks that @soubenz has added in this PR? https://github.com/zenml-io/zenml/pull/1402/files

They might fit in well in all the standard hooks. Just a small suggestion

File: src/zenml/integrations/openai/hooks/open_ai_failure_hook.py
Comment (strickvl) [reply src/zenml/integrations/openai/hooks/open_ai_failure_hook.py:47]: Ok so we'll have the full `rich` exception now instead of the standard Python one. Will make the change.

Comment (strickvl) [reply src/zenml/integrations/openai/hooks/open_ai_failure_hook.py:47]: Done

**Comment by strickvl** [comment]
> Apart from my one comment above, how about adding in these blocks that @soubenz has added in this PR? https://github.com/zenml-io/zenml/pull/1402/files
> 
> They might fit in well in all the standard hooks. Just a small suggestion

We can probably just wait until #1402 has been merged and then update as appropriate?

**Review by htahir1 — CHANGES_REQUESTED** [review]
Ah caught one more thing. Lets enforce that the `OPENAI_API_KEY` has to be set for these hooks?

Also we should merge in your other hooks fix PR before merging this one in to avoid conflcits (You have approvals so should be good to go the other one)

File: src/zenml/integrations/openai/hooks/open_ai_failure_hook.py
Comment (htahir1) [inline src/zenml/integrations/openai/hooks/open_ai_failure_hook.py:45]: do we need to check the alerter?

Comment (strickvl) [reply src/zenml/integrations/openai/hooks/open_ai_failure_hook.py:45]: If we don't check it and you don't have an alerter, later on when we run `context.stack.alerter.post(message)` this will raise an error.

**Review by htahir1 — APPROVED** [review]

**Pull request merged**
